File: repo/api/app/schemas/strategy.py
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, Field


class StrategyFeatureImportanceOut(BaseModel):
    feature: str
    importance: float | None = None
    stddev: float | None = None
    p_value: float | None = None
    n: int | None = None


class StrategyLeaderboardItemOut(BaseModel):
    model: str
    score_val: float | None = None
    fit_time: float | None = None
    pred_time_val: float | None = None


class StrategyRunOut(BaseModel):
    id: int
    strategy_code: str
    strategy_name: str
    as_of: date
    label_horizon: int
    status: str
    model_path: str | None = None
    train_rows: int
    scored_rows: int
    trained_at: datetime
    evaluation: dict[str, float | int | str | None] = Field(default_factory=dict)
    leaderboard: list[StrategyLeaderboardItemOut] = Field(default_factory=list)
    feature_importance: list[StrategyFeatureImportanceOut] = Field(default_factory=list)


class StrategyDriverOut(BaseModel):
    label: str
    tone: str
    value: float | None = None
    display_value: str | None = None


class StrategyFeatureValueOut(BaseModel):
    name: str
    value: float | None = None
    display_value: str | None = None


class SmokeButtCandidateOut(BaseModel):
    symbol: str
    name: str
    market: str
    sector: str
    as_of: date
    score: float
    rank: int
    percentile: float
    expected_return: float | None = None
    signal: str
    summary: str | None = None
    signal_explanation: str | None = None


class SmokeButtDetailOut(SmokeButtCandidateOut):
    run: StrategyRunOut
    drivers: list[StrategyDriverOut] = Field(default_factory=list)
    feature_values: list[StrategyFeatureValueOut] = Field(default_factory=list)


class SmokeButtPageOut(BaseModel):
    items: list[SmokeButtCandidateOut] = Field(default_factory=list)
    total: int
    limit: int
    offset: int
    run: StrategyRunOut | None = None


class SmokeButtTrainIn(BaseModel):
    as_of: date | None = None
    horizon_days: int = Field(60, ge=20, le=240)
    sample_step: int = Field(21, ge=5, le=63)
    time_limit_seconds: int | None = Field(120, ge=30, le=1800)
    force_retrain: bool = False


class SmokeButtTrainOut(BaseModel):
    run: StrategyRunOut
    items: list[SmokeButtCandidateOut] = Field(default_factory=list)


class StrategyBacktestCurvePointOut(BaseModel):
    date: date
    period_return: float | None = None
    cumulative_return: float | None = None


class StrategyBacktestBucketOut(BaseModel):
    bucket: str
    label: str
    bucket_index: int
    avg_return: float | None = None
    win_rate: float | None = None
    max_drawdown: float | None = None
    avg_predicted_return: float | None = None
    sample_count: int = 0
    period_count: int = 0
    curve: list[StrategyBacktestCurvePointOut] = Field(default_factory=list)


class StrategyBacktestWindowSummaryOut(BaseModel):
    top_bucket_return: float | None = None
    top_bucket_win_rate: float | None = None
    top_bucket_max_drawdown: float | None = None
    spread_return: float | None = None
    spread_win_rate: float | None = None
    spread_hit_rate: float | None = None
    monotonicity: float | None = None
    sample_count: int = 0
    period_count: int = 0


class StrategyBacktestWindowOut(BaseModel):
    horizon_days: int
    rebalance_step: int
    buckets: list[StrategyBacktestBucketOut] = Field(default_factory=list)
    summary: StrategyBacktestWindowSummaryOut


class SmokeButtBacktestConfidenceOut(BaseModel):
    validation_rank_ic: float | None = None
    validation_mae: float | None = None
    validation_rmse: float | None = None
    spread_return_20d: float | None = None
    spread_return_60d: float | None = None
    monotonicity_20d: float | None = None
    monotonicity_60d: float | None = None
    top_bucket_win_rate_20d: float | None = None
    top_bucket_win_rate_60d: float | None = None
    period_count_20d: int = 0
    period_count_60d: int = 0
    sample_count_20d: int = 0
    sample_count_60d: int = 0


class SmokeButtBacktestOut(BaseModel):
    run: StrategyRunOut
    market: str | None = None
    bucket_count: int
    windows: list[StrategyBacktestWindowOut] = Field(default_factory=list)
    confidence: SmokeButtBacktestConfidenceOut

File: repo/web/src/services/api/strategy.ts
import { ApiPage, ApiQueryOptions, request, requestJson } from "./core";

export type StrategySignal = "strong_buy" | "buy" | "watch" | "avoid";

export type SmokeButtFeatureImportance = {
  feature: string;
  importance?: number | null;
  stddev?: number | null;
  p_value?: number | null;
  n?: number | null;
};

export type SmokeButtLeaderboardItem = {
  model: string;
  score_val?: number | null;
  fit_time?: number | null;
  pred_time_val?: number | null;
};

export type SmokeButtRun = {
  id: number;
  strategy_code: string;
  strategy_name: string;
  as_of: string;
  label_horizon: number;
  status: string;
  model_path?: string | null;
  train_rows: number;
  scored_rows: number;
  trained_at: string;
  evaluation: Record<string, number | string | null>;
  leaderboard: SmokeButtLeaderboardItem[];
  feature_importance: SmokeButtFeatureImportance[];
};

export type SmokeButtCandidate = {
  symbol: string;
  name: string;
  market: string;
  sector: string;
  as_of: string;
  score: number;
  rank: number;
  percentile: number;
  expected_return?: number | null;
  signal: StrategySignal;
  summary?: string | null;
  signal_explanation?: string | null;
};

export type SmokeButtDriver = {
  label: string;
  tone: string;
  value?: number | null;
  display_value?: string | null;
};

export type SmokeButtFeatureValue = {
  name: string;
  value?: number | null;
  display_value?: string | null;
};

export type SmokeButtDetailResponse =
  | (SmokeButtCandidate & {
      run: SmokeButtRun;
      drivers: SmokeButtDriver[];
      feature_values: SmokeButtFeatureValue[];
    })
  | null;

export type SmokeButtListResponse = ApiPage<SmokeButtCandidate> & {
  run?: SmokeButtRun | null;
};

export type SmokeButtBacktestCurvePoint = {
  date: string;
  period_return?: number | null;
  cumulative_return?: number | null;
};

export type SmokeButtBacktestBucket = {
  bucket: string;
  label: string;
  bucket_index: number;
  avg_return?: number | null;
  win_rate?: number | null;
  max_drawdown?: number | null;
  avg_predicted_return?: number | null;
  sample_count: number;
  period_count: number;
  curve: SmokeButtBacktestCurvePoint[];
};

export type SmokeButtBacktestWindowSummary = {
  top_bucket_return?: number | null;
  top_bucket_win_rate?: number | null;
  top_bucket_max_drawdown?: number | null;
  spread_return?: number | null;
  spread_win_rate?: number | null;
  spread_hit_rate?: number | null;
  monotonicity?: number | null;
  sample_count: number;
  period_count: number;
};

export type SmokeButtBacktestWindow = {
  horizon_days: number;
  rebalance_step: number;
  buckets: SmokeButtBacktestBucket[];
  summary: SmokeButtBacktestWindowSummary;
};

export type SmokeButtBacktestConfidence = {
  validation_rank_ic?: number | null;
  validation_mae?: number | null;
  validation_rmse?: number | null;
  spread_return_20d?: number | null;
  spread_return_60d?: number | null;
  monotonicity_20d?: number | null;
  monotonicity_60d?: number | null;
  top_bucket_win_rate_20d?: number | null;
  top_bucket_win_rate_60d?: number | null;
  period_count_20d: number;
  period_count_60d: number;
  sample_count_20d: number;
  sample_count_60d: number;
};

export type SmokeButtBacktestResponse = {
  run: SmokeButtRun;
  market?: "A" | "HK" | "US" | null;
  bucket_count: number;
  windows: SmokeButtBacktestWindow[];
  confidence: SmokeButtBacktestConfidence;
} | null;

export async function getSmokeButtStrategyLeaderboard(
  params?: {
    market?: "A" | "HK" | "US";
    signal?: StrategySignal;
    limit?: number;
    offset?: number;
  },
  options?: ApiQueryOptions,
) {
  const query = new URLSearchParams();
  if (params?.market) query.set("market", params.market);
  if (params?.signal) query.set("signal", params.signal);
  if (params?.limit !== undefined) query.set("limit", String(params.limit));
  if (params?.offset !== undefined) query.set("offset", String(params.offset));
  const suffix = query.toString();
  return request<SmokeButtListResponse>(`/api/strategy/smoke-butt${suffix ? `?${suffix}` : ""}`, {
    ...options,
    label: options?.label ?? "strategy-score-leaderboard",
  });
}

export async function getSmokeButtStrategyDetail(symbol: string, options?: ApiQueryOptions) {
  return request<SmokeButtDetailResponse>(`/api/strategy/smoke-butt/${encodeURIComponent(symbol)}`, {
    ...options,
    label: options?.label ?? "strategy-score-detail",
  });
}

export async function getSmokeButtBacktest(
  params?: {
    market?: "A" | "HK" | "US";
    bucket_count?: number;
  },
  options?: ApiQueryOptions,
) {
  const query = new URLSearchParams();
  if (params?.market) query.set("market", params.market);
  if (params?.bucket_count !== undefined) query.set("bucket_count", String(params.bucket_count));
  const suffix = query.toString();
  return request<SmokeButtBacktestResponse>(`/api/strategy/smoke-butt/backtest${suffix ? `?${suffix}` : ""}`, {
    ...options,
    label: options?.label ?? "strategy-score-backtest",
  });
}

export async function trainSmokeButtStrategy(payload?: {
  as_of?: string;
  horizon_days?: number;
  sample_step?: number;
  time_limit_seconds?: number;
  force_retrain?: boolean;
}) {
  return requestJson<{ run: SmokeButtRun; items: SmokeButtCandidate[] }>(
    "/api/strategy/smoke-butt/train",
    {
      as_of: payload?.as_of ?? null,
      horizon_days: payload?.horizon_days ?? 60,
      sample_step: payload?.sample_step ?? 21,
      time_limit_seconds: payload?.time_limit_seconds ?? 120,
      force_retrain: payload?.force_retrain ?? false,
    },
    { retry: 0, label: "strategy-score-train" },
  );
}
